### PR TITLE
Add event pipeline and dispatcher

### PR DESCRIPTION
**Pull request by kirodis**

### CODE REVIEW

File: eventstore/src/main/java/com/github/yafna/events/dispatcher/EventDispatcher.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ * 2. Notify pipelines about stored events
+ * 3. Manage transactions if event store behind it supports them
+ */
+@AllArgsConstructor
+public abstract class EventDispatcher {
+
+    private final EventStore store;
+    private final ForkJoinPool executor;
+
+    public <T> Event store(String origin, String aggregateId, T event) {
```
Comment (yafna) [inline eventstore/src/main/java/com/github/yafna/events/dispatcher/EventDispatcher.java:37]: no usages in code

File: eventstore/src/main/java/com/github/yafna/events/pipelines/EventPipeline.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    public EventPipeline(
+            EventDispatcher dispatcher, Class<T> eventType, EventHandler<T> handler, String handlerId, Clock clock, ForkJoinPool executor
+    ) {
+        origin = PayloadUtils.origin(eventType);
+        this.executor = executor;
+        this.dispatcher = dispatcher;
+        this.handler = handler;
+        this.handlerId = handlerId;
+        Duration timeWindow = Duration.ofDays(1);
+        Instant since = clock.instant().minus(timeWindow);
```
Comment (yafna) [inline eventstore/src/main/java/com/github/yafna/events/pipelines/EventPipeline.java:42]: why 1 day?

File: eventstore/src/main/java/com/github/yafna/events/pipelines/EventPipeline.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                        event -> process(event.getMeta(), event.getPayload())
+                ).max(Comparator.naturalOrder()).orElseThrow(() -> {
+                    log.error("Got empty stream for [{}].[{}]", origin, type);
+                    return new IllegalStateException("Should never return empty stream");
+                });
+            }
+        }
+    }
+
+    private Instant process(Event meta, T payload) {
```
Comment (yafna) [inline eventstore/src/main/java/com/github/yafna/events/pipelines/EventPipeline.java:64]: some comment would be useful. for example why "handlers" word is hardcoded

File: eventstore/src/test/groovy/com/github/yafna/events/pipelines/EventPipelineSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    ForkJoinPool executor = new ForkJoinPool()
+    Clock clock = Clock.systemUTC()
+    GsonFileEventStore store = new GsonFileEventStore(clock, File.createTempDir())
+
+    def 'given [rabbit.added] event persisted should call EventHandler for it'() {
+        given:
+            def handler = new RabbitCountingHandler()
+            EventPipeline<Rabbit, RabbitAdded> subj = new EventPipeline(
+                    new GsonEventDispatcher(store, executor), RabbitAdded.class, handler, "rabbit-counter", clock, executor 
+            )
```
Comment (yafna) [inline eventstore/src/test/groovy/com/github/yafna/events/pipelines/EventPipelineSpec.groovy:29]: first test: handler.count.get == 0

**Review by yafna — COMMENTED** [review]
see the comments

File: eventstore/src/test/groovy/com/github/yafna/events/pipelines/EventPipelineSpec.groovy (hunk 4 above)
Comment (kirodis) [reply eventstore/src/test/groovy/com/github/yafna/events/pipelines/EventPipelineSpec.groovy:29]: The handler is created in this test and the handlers is test code. Such check would be excessive

File: eventstore/src/main/java/com/github/yafna/events/pipelines/EventPipeline.java (hunk 2 above)
Comment (kirodis) [reply eventstore/src/main/java/com/github/yafna/events/pipelines/EventPipeline.java:42]: A default value for time window. Has to be made a constructor parameter along with adding some tests

File: eventstore/src/main/java/com/github/yafna/events/dispatcher/EventDispatcher.java (hunk 1 above)
Comment (kirodis) [reply eventstore/src/main/java/com/github/yafna/events/dispatcher/EventDispatcher.java:37]: Fixed

**Pull request merged**